app/scripts/modules/core/src/function/function.read.service.ts

```diff
@@ -0,0 +1,58 @@
+import { IPromise, IQService, module } from 'angular';
+
+import { API } from 'core/api/ApiService';
+import { IFunctionSourceData, IFunction } from 'core/domain';
+
+export interface IFunctionByAccount {
+  name: string;
+  accounts: Array<{
+    name: string;
+    regions: Array<{
+      name: string;
+      functions: IFunctionSourceData[];
+    }>;
+  }>;
+}
+
+export class FunctionReader {
+  public static $inject = ['$q', 'functionTransformer'];
+  public constructor(private $q: IQService, private functionTransformer: any) {}
+
+  public loadFunctions(applicationName: string): IPromise<IFunctionSourceData[]> {
+    return API.one('applications', applicationName)
+      .all('functions')
+      .getList()
+      .then((functions: IFunctionSourceData[]) => {
+        functions = this.functionTransformer.normalizeFunctionSet(functions);
+        return this.$q.all(functions.map(fn => this.normalizeFunction(fn)));
+      });
+  }
+
+  public getFunctionDetails(
+    cloudProvider: string,
+    account: string,
+    region: string,
+    name: string,
+  ): IPromise<IFunctionSourceData[]> {
+    return API.all('functions')
+      .withParams({ provider: cloudProvider, functionName: name, region: region, account: account })
+      .get();
+  }
+
+  public listFunctions(cloudProvider: string): IPromise<IFunctionByAccount[]> {
+    return API.all('functions')
+      .withParams({ provider: cloudProvider })
+      .getList();
+  }
+
+  private normalizeFunction(functionDef: IFunctionSourceData): IPromise<IFunction> {
+    return this.functionTransformer.normalizeFunction(functionDef).then((fn: IFunction) => {
+      fn.cloudProvider = fn.cloudProvider || 'aws';
```

Comment:
Does the backend ever return functions without a cloudProvider field?  This seems a little odd since I assume the backend always returns 'aws' for Lambdas?